### PR TITLE
[BugFix] fix compression bugs

### PR DESCRIPTION
**Pull request by J-shang**
### Description ###
Depend on #5138

Fixed:
- speedup rand_like_with_shape is easy to overflow when dtype=torch.int8.
- https://github.com/microsoft/nni/issues/5037
- https://github.com/microsoft/nni/issues/4818

#### Test Options ####
  - [ ] fast test
  - [ ] full test - HPO
  - [ ] full test - NAS
  - [ ] full test - compression

### Checklist ###
  - [ ] test case
  - [ ] doc

### CODE REVIEW

File: nni/compression/pytorch/utils/shape_dependency.py [hunk 1]
```diff
@@ -152,8 +152,12 @@ def _get_parent_layers(self, node):
         parent_layers = []
         queue = []
         queue.append(node)
+        viewed_set = set()
```
Comment (Louis-J) [inline nni/compression/pytorch/utils/shape_dependency.py:155]: suggest: other variables to record visited nodes are named 'visited'

File: nni_assets/compression/mnist_model.py [hunk 2]
```diff
@@ -38,13 +36,13 @@ def forward(self, x):
 from torchvision import datasets, transforms
 
 train_loader = torch.utils.data.DataLoader(
-    datasets.MNIST(root_path / 'data', train=True, download=True, transform=transforms.Compose([
+    datasets.MNIST('data', train=True, download=True, transform=transforms.Compose([
         transforms.ToTensor(),
         transforms.Normalize((0.1307,), (0.3081,))
```
Comment (Louis-J) [inline nni_assets/compression/mnist_model.py:41]: are these magic number?

Comment (J-shang) [reply nni_assets/compression/mnist_model.py:41]: yes, magic number for mnist dataset

File: nni/compression/pytorch/utils/shape_dependency.py (hunk 1 above)
Comment (J-shang) [reply nni/compression/pytorch/utils/shape_dependency.py]: fixed

**Review by QuanluZhang — APPROVED** [review]

**Pull request merged**